### PR TITLE
[Reviewer: Matt] Chronos scale docs

### PR DESCRIPTION
**Pull request by eleanor-merry**
Matt, can you please review these docs for Chronos scaling

### CODE REVIEW

File: doc/api.md [hunk 1]
```diff
@@ -100,14 +97,91 @@ _Note: Performing a PUT to a non-existent timer will create a timer with the spe
...
+
+#### Request (GET)
+
+    GET /timers
+
+This URL requests information about timers that are on the receiving node that the requesting node will be a replica for under new cluster configuration. 
+ 
+It takes three mandatory parameters
+
+* `requesting-node=<address>` - The address of the requesting node. This must match a node in the Chronos cluster
```
Comment (mirw) [inline doc/api.md]: Why is the requesting node relevant?  Surely the important parameter is the address of the node that should own timers?

File: doc/api.md [hunk 2]
```diff
@@ -100,14 +97,91 @@ _Note: Performing a PUT to a non-existent timer will create a timer with the spe
...
+* `sync-mode=<sync-mode>` - The synchronization mode. The only currently supported value is SCALE.
```
Comment (mirw) [inline doc/api.md]: Why does this matter?

File: doc/api.md [hunk 3]
```diff
@@ -100,14 +97,91 @@ _Note: Performing a PUT to a non-existent timer will create a timer with the spe
...
+
+    Range: 100
+
+#### Response (GET)
+
+If the GET is missing any of the mandatory parameters or any of the mandatory parameters are invalid then the request is rejected with the following result codes:
+
+* Any of the mandatory parameters are missing - `400 Bad Request`
+* The requesting-node address doesn't correspond to a Chronos node in the cluster - `404 Not Found`
+* The sync-mode requested is isn't a supported mode - `400 Bad Request`
```
Comment (mirw) [inline doc/api.md]: We really should help the user here - don't just return 400 bad request for everything - try to give them an error message that they can use.

File: doc/api.md [hunk 4]
```diff
@@ -100,14 +97,91 @@ _Note: Performing a PUT to a non-existent timer will create a timer with the spe
...
+* The cluster-view-id supplied by the requesting node doesn't match the receiving node's view of the cluster ID - `400 Bad Request`
+
+The response to a valid GET request is a `200 OK` or a `206 Partial Content`, with a JSON body containing timer information. A `206` is sent instead of a `200` if there are more timers that could be sent, but aren't because of the maximum timer limit. The 206 response includes a `Content-Range` header that shows how many timers were included (e.g. `Content-Range: 100`)
```
Comment (mirw) [inline doc/api.md]: Is it valid to send a `Range: 100` and get back a 206 with `Content-Range: 99`?

File: doc/api.md [hunk 5]
```diff
@@ -100,14 +97,91 @@ _Note: Performing a PUT to a non-existent timer will create a timer with the spe
...
+                                      "sequence-number": <int>
+                                      "interval": <ms>,
+                                      "repeat-for": <ms>
+                                     },
+                           "callback": {"http": {"uri": <callback-uri>,
+                                                 "opaque": <opaque-data>
+                                                }
+                                       },
+                           "reliability": {"cluster-view-id": <cluster-view-id>,
+                                           "replicas": ["replica_1", ...]
```
Comment (mirw) [inline doc/api.md]: Why is the new replica list replicated?  Shouldn't the cluster agree on this?

File: doc/api.md [hunk 6]
```diff
@@ -100,14 +97,91 @@ _Note: Performing a PUT to a non-existent timer will create a timer with the spe
...
+* Any of the mandatory parameters are missing - `400 Bad Request`
+* The requesting-node address doesn't correspond to a Chronos node in the cluster - `404 Not Found`
+* The sync-mode requested is isn't a supported mode - `400 Bad Request`
+* The cluster-view-id supplied by the requesting node doesn't match the receiving node's view of the cluster ID - `400 Bad Request`
+
+The response to a valid GET request is a `200 OK` or a `206 Partial Content`, with a JSON body containing timer information. A `206` is sent instead of a `200` if there are more timers that could be sent, but aren't because of the maximum timer limit. The 206 response includes a `Content-Range` header that shows how many timers were included (e.g. `Content-Range: 100`)
+
+The JSON body in the response has the format:
+
+    {"timers": [{"Timer ID": id,
```
Comment (mirw) [inline doc/api.md]: Why are we using spaces in identifiers?  Given that this is an API, surely it's just going to cause developers headaches?  ...and shouldn't we be consistent about capitalization?

File: doc/api.md [hunk 7]
```diff
@@ -100,14 +97,91 @@ _Note: Performing a PUT to a non-existent timer will create a timer with the spe
...
+
+The DELETE body consists of the IDs of all the timers the node has just processed, paired with the replica number of the node for each timer. It has the format:
+
+    {"IDs": [{"ID": id1, "replica index": replica_index},
+             {"ID": id2, "replica index": replica_index}
+             ...
+            ]
+    }
+
+The `replica_index` is the index of the timer in the replica list (where 0 represents the primary). The `ID` is the timer ID.
```
Comment (mirw) [inline doc/api.md]: Be consistent about `replica_index` vs `replica index` (and using hyphens, e.g. `cluster-view-id`).

File: doc/clustering.md [hunk 8]
```diff
@@ -4,10 +4,6 @@ Chronos is a distributed timer service, based on an arbitrary-size cluster of co
 
 Chronos's cluster settings are controlled by `/etc/chronos/chronos.conf`, which includes:
 
-    [http]
-    bind-address = 1.2.3.4
-    bind-port = 7253
-
```
Comment (mirw) [inline doc/clustering.md]: Is this no longer valid?  Why not?

File: doc/configuration.md [hunk 9]
```diff
@@ -6,10 +6,15 @@ Chronos's configuration is set up in `/etc/chronos/chronos.conf`, and takes the
...
+    
     [cluster]
     localhost = 1.2.3.4            # The address of the local host
-    node = 1.2.3.4                 # The addresses of a node in the cluster
+    node = 1.2.3.4                 # The addresses of all nodes in the cluster. If the 
+    node = 1.2.3.5                 # address doesn't include a port, then the default-bind-port
+    node = 1.2.3.6                 # value is used. 
+    leaving = 2.3.4.5              # The addresses of all nodes that are leaving
+    leaving = 2.3.4.6              # the cluster (only used during scale-down)
+    default-bind-port = 7253       # Default port to connect to other Chronos servers on
```
Comment (mirw) [inline doc/configuration.md]: Is `default-bind-port` the right name?  Surely this is a port that we *connect* to?

File: doc/design/resynchronization.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+This document describes the resynchronization process in more detail.
+
+## Handling a resynchronization GET request
+
+When a node receives a GET request as part of resynchronization (see [here](../api.md#request-get) for the API) it does the following processing:
+
+The node loops through their timer wheel. For each timer it does the following processing
+
+* It compares the `cluster-view-id` on the GET request to the stored `cluster-view-id` in the timer(s). If these are the same, then the nodes stops checking the timer and moves onto the next timer.
```
Comment (mirw) [inline doc/design/resynchronization.md]: Context!

File: doc/scaling.md [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+the deterministic process would need to know what the cluster looked like when the timer was first
+created so it could repeat the algorithm to determine the correct replica list.
+
+Chronos's solution involves encoding the list of replicas into the ID returned to the user when they
+create/update a timer. The user then uses this ID when they want to update/delete the timer later. 
+ID creation and processing is described in more detail [here](design/hashing.md).
+
+## Rebalancing timers
+
+When a Chronos cluster scales up/down then we want to rebalance the timers, so that new Chronos nodes
```
Comment (mirw) [inline doc/scaling.md]: Define "rebalance timers" before using it.

File: doc/scaling.md [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+     * Node C updates its local store.
+
+ * Node D sees an update request
+     * Node D does not recognize this timer so it creates a record for it
+
+In this way, the timers will automatically spread themselves out over the larger cluster.
+
+### Rebalancing timers - active method
+
+The active resynchronization process has three steps:
```
Comment (mirw) [inline doc/scaling.md]: Context/structure!

**Comment by mirw** [comment]
Comments interspersed.

File: doc/clustering.md (hunk 8 above)
Comment (eleanor-merry) [inline doc/clustering.md]: It is still valid, I just didn't think it was particularly useful information for the clustering doc so I removed it.

File: doc/configuration.md (hunk 9 above)
Comment (eleanor-merry) [inline doc/configuration.md]: Agree- I'll change this

File: doc/clustering.md (hunk 8 above)
Comment (mirw) [inline doc/clustering.md]: OK.

File: doc/configuration.md (hunk 9 above)
Comment (bossmc) [inline doc/configuration.md]: Why is this not `bind-port`?  Bear in mind that you (used to be able to) can specify `node` as `<ip>:<port>` if you want.

File: doc/scaling.md (hunk 12 above)
Comment (bossmc) [inline doc/scaling.md]: :grinning:

**Pull request merged**